src/object/srv_obj_migrate.c

```diff
@@ -2070,6 +2119,150 @@ struct migrate_obj_val {
+ *
+ * Must be called on the appropriate VOS target xstream!
+ */
+static int
+destroy_existing_obj(struct migrate_pool_tls *tls, unsigned int tgt_idx,
+		     daos_unit_oid_t *oid, uuid_t cont_uuid)
+{
+	d_list_t *link;
+	int rc;
+
+	/* NOTE:
+	 * Note that only the object ID + shard is stored in the hash table, not
+	 * the padding that is also included in the daos_unit_oid_t struct
+	 * There will be many objects inserted in this hash table and so as
+	 * little data should be stored there as possible
+	 */
+	link = d_hash_rec_find(&tls->mpt_del_tabs[tgt_idx], oid,
+			       sizeof(daos_obj_id_t) + sizeof(uint32_t));
+	if (!link) {
+		/* Not actually storing anything in the table - just using it
+		 * to test set membership. The link stored is just the simplest
+		 * base list type
+		 */
+		struct migrate_del_obj_key *key;
+		struct ds_cont_child *cont;
+
+		rc = ds_cont_child_open_create(tls->mpt_pool_uuid,
+					       cont_uuid,
+					       &cont);
+		if (rc) {
+			D_WARN("Failed to open cont prior to migration; pool="
+			       DF_UUID" cont="DF_UUID"\n",
+			       DP_UUID(tls->mpt_pool_uuid), DP_UUID(cont_uuid));
+			/* This failure is not fatal - the object can't be
+			 * deleted, but it probably doesn't need to be if
+			 * the container can't even be opened. Proceed onwards,
+			 * marking the object as being ready to start migration
+			 */
+			D_GOTO(mark_obj_deleted, rc);
+		}
+
+		rc = vos_obj_delete(cont->sc_hdl, *oid);
+		if (rc != 0)
+			D_ERROR("Migrate failed to destroy object prior to "
+				"reintegration: pool/object "DF_UUID"/"DF_UOID
+				" rc: "DF_RC"\n",
+				DP_UUID(tls->mpt_pool_uuid), DP_UOID(*oid),
+				DP_RC(rc));
+			/*
+			 * Despite this error, continue anyway to add this key
+			 * to the table. Attempting to delete this object again
+			 * later will only make things worse
+			 */
+		else
+			D_DEBUG(DB_REBUILD,
+				"destroyed pool/object "DF_UUID"/"DF_UOID
+				" before reintegration\n",
+				DP_UUID(tls->mpt_pool_uuid), DP_UOID(*oid));
+
+		ds_cont_child_put(cont);
+
+mark_obj_deleted:
+
+		/* Insert a link into the hash table to mark this cont_uuid as
+		 * having already been initialized
+		 */
+		D_ALLOC_PTR(key);
+		if (key == NULL)
+			return -DER_NOMEM;
+
+		key->id_pub.hi = oid->id_pub.hi;
+		key->id_pub.lo = oid->id_pub.lo;
+		key->id_shard = oid->id_shard;
+
+		D_INIT_LIST_HEAD(&key->obj_link);
+		rc = d_hash_rec_insert(&tls->mpt_del_tabs[tgt_idx], key,
+				       sizeof(daos_obj_id_t) + sizeof(uint32_t),
+				       &key->obj_link, true);
+		if (rc) {
+			D_ERROR("Failed to insert uuid table entry "DF_RC"\n",
+				DP_RC(rc));
+			D_FREE(key);
+			return rc;
+		}
+	}
+
+	return DER_SUCCESS;
+}
+
+/* This ULT runs on the target where the data will ultimately be migrated
+ * to do any needed preparation work prior to actually doing object migration
+ *
+ * Reintegration uses this to delete objects prior to migrating them
+ */
+static void
+migrate_one_obj_prepare_ult(void *data)
```

Comment:
this should work... but I also noticed your message in the channel and said it could be also in migrate_one_ult()? I'm not asking for a change, just wondering why it's in this way.

Reply:
Changed in the new version to remove this intermediate ULT. The migrate_one_ult() is scheduled on a pseudorandom ULT so wasn't suitable for doing object deletes. New patch runs that ULT on the specific target ID in the case of reintegration, and otherwise schedules pseudorandomly to get the best of both worlds.